lucet-runtime/lucet-runtime-internals/src/instance/execution.rs

```diff
@@ -125,6 +127,41 @@ pub struct KillState {
     tid_change_notifier: Condvar,
 }
 
+/// This function is called by `lucet_context_activate` before entering a region of guest code.
+///
+/// This function is responsible for setting the terminable flag on the instance's
+/// [`KillState`](struct.KillState.html), so that we can appropriately signal the instance to
+/// terminate. If a pending instance was signalled before it started executing guest code, we will
+/// terminate the instance and swap back to its host context.
+///
+///
+/// entering a guest region, so that we can use signals to terminate the instance.  check that the
+/// instance has not already been remotely terminated by a killswitch.  set the `terminable` flag
+/// on our instance's [`KillState`](struct.KillState.html).
+///
+/// # Safety
+///
+/// This function will call [Instance::terminate](struct.Instance.html#method.terminate) if the
+/// execution domain is [`Domain::Cancelled`](enum.Domain.html). In that case, this function will
+/// not return, and we will swap back to the host context without unwinding.
+pub unsafe extern "C" fn enter_guest_region(instance: *mut Instance) {
+    let mut current_domain = (*instance).kill_state.execution_domain.lock().unwrap();
+    match *current_domain {
+        // We are about to start executing a guest. Set the execution domain accordingly.
+        Domain::Pending => *current_domain = Domain::Guest,
+        // We are about to enter guest code for an instance that has already been terminated.
+        // Instead, we should forgo entering the guest region entirely, terminate the instance, and
+        // swap back to the host without any unwinding.
+        Domain::Cancelled => (*instance).terminate(TerminationDetails::Remote),
```

Comment:
This function has some strong warnings in its comments ("_This is almost certainly not what you want to use to terminate from a hostcall_") so I wanted to make sure this wasn't a total error. In this case though, we don't want to unwind because there isn't anything _wound_!

https://github.com/bytecodealliance/lucet/blob/ee003cd27643a4ad5a745cd3a4d02d5315d3a8b2/lucet-runtime/lucet-runtime-internals/src/vmctx.rs#L425-L429

Reply:
I think "_almost certainly not what you want to use_" is more a warning to hostcall implementors because resources like Mutex guards wouldn't get dropped, possibly leading to deadlocks in the host after exiting. Because no custom user code has run at this point, guest or host, this should be fine.

Reply:
Figured that was the case, but those were strong enough words that I wanted to be sure :slightly_smiling_face: Sounds good!